File: conformance/independent-python/verify.py
#!/usr/bin/env python3
"""Independent verifier for ĀML Decision Core 1.

This intentionally does not import aml-core. It implements the published
Decision Core 1 rule directly so the vectors can test cross-implementation
agreement rather than a wrapper around the reference runtime.
"""

import json
import math
from pathlib import Path
import sys

ROOT = Path(__file__).resolve().parents[1]
VECTORS = ROOT / "decision-core-1.json"
INVALID_VECTORS = ROOT / "decision-core-1-invalid.json"
PROTOCOL = "aml-conformance/decision-core-1"


def decision(attention_cost, restoration_value):
    if isinstance(attention_cost, bool) or isinstance(restoration_value, bool):
        raise TypeError("scores must be JSON numbers, not booleans")
    if not isinstance(attention_cost, (int, float)) or not isinstance(restoration_value, (int, float)):
        raise TypeError("scores must be JSON numbers")
    if not math.isfinite(attention_cost) or not math.isfinite(restoration_value):
        raise ValueError("scores must be finite")
    return "ALLOW" if restoration_value >= attention_cost else "SUPPRESS"


def load(path):
    payload = json.loads(path.read_text(encoding="utf-8"))
    if payload.get("protocol") != PROTOCOL:
        raise ValueError(f"unexpected conformance protocol in {path}")
    return payload


def main():
    try:
        payload = load(VECTORS)
        invalid_payload = load(INVALID_VECTORS)
    except Exception as exc:
        print(f"FAIL: {exc}", file=sys.stderr)
        return 2

    failures = []

    for vector in payload["vectors"]:
        try:
            actual = decision(vector["attention_cost"], vector["restoration_value"])
        except Exception as exc:
            failures.append((vector["id"], vector["expected_decision"], f"ERROR:{type(exc).__name__}"))
            continue
        expected = vector["expected_decision"]
        if actual != expected:
            failures.append((vector["id"], expected, actual))

    rejected = 0
    for vector in invalid_payload["vectors"]:
        try:
            decision(vector["attention_cost"], vector["restoration_value"])
        except (KeyError, TypeError, ValueError):
            rejected += 1
        else:
            failures.append((vector["id"], "REJECT", "ACCEPT"))

    if failures:
        for vector_id, expected, actual in failures:
            print(f"FAIL {vector_id}: expected {expected}, got {actual}", file=sys.stderr)
        return 1

    valid_count = len(payload["vectors"])
    invalid_count = len(invalid_payload["vectors"])
    print(f"PASS {PROTOCOL} ({valid_count}/{valid_count} decisions; {rejected}/{invalid_count} invalid rejected)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: conformance/independent-javascript/verify.mjs
#!/usr/bin/env node
/**
 * Independent verifier for ĀML Decision Core 1.
 *
 * This file intentionally imports no aml-core modules. It implements the
 * published conformance rule directly from CONFORMANCE.md and the vectors.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const here = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(here, "..");
const protocol = "aml-conformance/decision-core-1";

function load(name) {
  const payload = JSON.parse(fs.readFileSync(path.resolve(root, name), "utf8"));
  if (payload.protocol !== protocol) {
    throw new Error(`unexpected conformance protocol in ${name}`);
  }
  return payload;
}

function decision(attentionCost, restorationValue) {
  if (typeof attentionCost !== "number" || typeof restorationValue !== "number") {
    throw new TypeError("scores must be JSON numbers");
  }
  if (!Number.isFinite(attentionCost) || !Number.isFinite(restorationValue)) {
    throw new TypeError("scores must be finite JSON numbers");
  }
  return restorationValue >= attentionCost ? "ALLOW" : "SUPPRESS";
}

let payload;
let invalidPayload;
try {
  payload = load("decision-core-1.json");
  invalidPayload = load("decision-core-1-invalid.json");
} catch (error) {
  console.error(`FAIL: ${error.message}`);
  process.exit(2);
}

const failures = [];
for (const vector of payload.vectors) {
  try {
    const actual = decision(vector.attention_cost, vector.restoration_value);
    if (actual !== vector.expected_decision) {
      failures.push({ id: vector.id, expected: vector.expected_decision, actual });
    }
  } catch (error) {
    failures.push({ id: vector.id, expected: vector.expected_decision, actual: `ERROR:${error.name}` });
  }
}

let rejected = 0;
for (const vector of invalidPayload.vectors) {
  try {
    decision(vector.attention_cost, vector.restoration_value);
    failures.push({ id: vector.id, expected: "REJECT", actual: "ACCEPT" });
  } catch (error) {
    if (error instanceof TypeError) {
      rejected += 1;
    } else {
      failures.push({ id: vector.id, expected: "REJECT", actual: `ERROR:${error.name}` });
    }
  }
}

if (failures.length) {
  for (const failure of failures) {
    console.error(`FAIL ${failure.id}: expected ${failure.expected}, got ${failure.actual}`);
  }
  process.exit(1);
}

const validCount = payload.vectors.length;
const invalidCount = invalidPayload.vectors.length;
console.log(`PASS ${protocol} (${validCount}/${validCount} decisions; ${rejected}/${invalidCount} invalid rejected)`);
